src/GroverSimulator.cpp

```diff
@@ -5,89 +5,93 @@
-        qc_diffusion.emplace_back<qc::StandardOperation>(n_qubits + n_anciallae, i, qc::H);
+    for (dd::QubitCount i = 0; i < n_qubits; ++i) {
+        qc_diffusion.h(i);
     }
-    //qc_diffusion.print();
-    dd::Edge diffusion_op{dd::buildFunctionality(&qc_diffusion, dd)};
 
-    dd::Edge full_iteration{dd->multiply(oracle_op, diffusion_op)};
-    dd->incRef(full_iteration);
+    dd::Edge diffusion_op{dd::buildFunctionality(&qc_diffusion, this->dd)};
 
-    rootEdge = dd->makeZeroState(n_qubits + n_anciallae);
-    rootEdge = dd->multiply(setup_op, rootEdge);
-    dd->incRef(rootEdge);
+    dd::Edge full_iteration{this->dd->multiply(oracle_op, diffusion_op)};
+    this->dd->incRef(full_iteration);
 
-    unsigned int j_pre = 0;
+    assert(n_qubits + n_anciallae <= std::numeric_limits<dd::QubitCount>::max());
+    this->rootEdge = this->dd->makeZeroState(static_cast<dd::QubitCount>(n_qubits + n_anciallae));
+    this->rootEdge = this->dd->multiply(setup_op, this->rootEdge);
+    this->dd->incRef(this->rootEdge);
+
+    std::size_t j_pre = 0;
 
     while ((iterations - j_pre) % 8 != 0) {
         //std::clog << "[INFO]  Pre-Iteration " << j_pre+1 << " of " << iterations%8 << " -- size:" << dd->size(rootEdge)  << "\n";
-        auto tmp = dd->multiply(full_iteration, rootEdge);
-        dd->incRef(tmp);
-        dd->decRef(rootEdge);
-        rootEdge = tmp;
-        dd->garbageCollect();
+        auto tmp = this->dd->multiply(full_iteration, this->rootEdge);
+        this->dd->incRef(tmp);
+        this->dd->decRef(this->rootEdge);
+        this->rootEdge = tmp;
+        this->dd->garbageCollect();
         j_pre++;
     }
 
-    for (unsigned long long j = j_pre; j < iterations; j += 8) {
+    for (std::size_t j = j_pre; j < iterations; j += 8) {
         //std::clog << "[INFO]  Iteration " << j+1 << " of " << iterations << " -- size:" << dd->size(rootEdge)  << "\n";
-        auto tmp = dd->multiply(full_iteration, rootEdge);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        tmp      = dd->multiply(full_iteration, tmp);
-        dd->incRef(tmp);
-        dd->decRef(rootEdge);
-        rootEdge = tmp;
-        dd->garbageCollect();
+        auto tmp = this->dd->multiply(full_iteration, this->rootEdge);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
+        tmp      = this->dd->multiply(full_iteration, tmp);
```

Comment:
this looks kinda ugly tbh. could you maybe refactor that to a loop 🙃

Reply:
This is explicitly unrolled as empirically eight multiplcations before garbage collection gave the best performance.

Reply:
Yeah, got that. Still would be way less ugly as
```suggestion
        auto tmp = this->dd->multiply(full_iteration, this->rootEdge);
        for (std::size_t i=0; i<7; ++i) {
          tmp      = this->dd->multiply(full_iteration, tmp);
        }
```